substrate/client/consensus/beefy/src/communication/gossip.rs

```diff
@@ -385,9 +420,28 @@ where
 		mut data: &[u8],
 	) -> ValidationResult<B::Hash> {
 		let raw = data;
+		// Have fisherman check the vote or proof for fork equivocations regardless of
+		// the action on it:
+		// 1. We check votes/proofs on past rounds and active rounds since they might equivocate
+		//    against grandpa-finalized state.
+		// 2. We check votes/proofs on future non-active rounds since these should not have been
+		//    voted on yet (from the client's perspective). In case the block is not even in the
+		//    client's history, but is in fact already finalized, the resulting equivocation report
+		//    will be ineffectual.
+		// The check is best-effort and ignores errors such as state pruned. Accepted votes are also
+		// checked against vote equivocations in
+		// `sc_consensus_beefy::worker::BeefyWorker::handle_vote`
 		let action = match GossipMessage::<B>::decode_all(&mut data) {
-			Ok(GossipMessage::Vote(msg)) => self.validate_vote(msg, sender),
-			Ok(GossipMessage::FinalityProof(proof)) => self.validate_finality_proof(proof, sender),
+			Ok(GossipMessage::Vote(msg)) => {
+				let action = self.validate_vote(&msg, sender);
+				let _ = self.fisherman.check_vote(msg);
```

Comment:
If we do this, I don't know if we also need to report vote equivocations in `BeefyWorker::handle_vote`. From what I understand, this way we report vote equivocations twice. Once here using a `ForkEquivocationProof`, and once in `BeefyWorker::handle_vote` using `VoteEquivocationProof`.

Reply:
Also not sure if we still need the `VoteEquivocationProof` enum variant.

Reply:
Actually this might be ok. But I think we should define some equivocation reporting primitives (generating ownership proofs, reporting) somewhere so that they can be used both by the worker for reporting vote equivocations and by the fisherman.

Reply:
> But I think we should define some equivocation reporting primitives (generating ownership proofs, reporting) somewhere so that they can be used both by the worker for reporting vote equivocations and by the fisherman.

Opened a separate PR for this: https://github.com/paritytech/polkadot-sdk/pull/4328. Also it adds a basic `Fisherman` struct and renames `EquivocationProof` to `DoubleVotingProof` in order to make this PR smaller and merge parts of it incrementally.